libbpf-tools/cpudist.c

```diff
@@ -0,0 +1,390 @@
+				argp_usage(state);
+			}
+			env.times = times;
+		} else {
+			fprintf(stderr,
+				"unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		pos_args++;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static int bump_memlock_rlimit(void)
+{
+	struct rlimit rlim_new = {
+		.rlim_cur	= RLIM_INFINITY,
+		.rlim_max	= RLIM_INFINITY,
+	};
+
+	return setrlimit(RLIMIT_MEMLOCK, &rlim_new);
+}
+
+static int get_pid_max(void)
+{
+	int pid_max;
+	FILE *f;
+
+	f = fopen("/proc/sys/kernel/pid_max", "r");
+	if (!f)
+		return -1;
+	if (fscanf(f, "%d\n", &pid_max) != 1)
+		return -1;
+	return pid_max;
+}
+
+static void sig_handler(int sig)
+{
+	exiting = true;
+}
+
+static void print_global_log2_hist(int fd)
+{
+	__u32 lookup_key = -1, next_key, i = 0, *keys;
+	DECLARE_LIBBPF_OPTS(bpf_map_batch_opts, opts,
+		.elem_flags = 0,
+		.flags = 0,
+	);
+	__u64 *vals, *zeros, val;
+	int max_slots = 65, err;
+
+	vals = calloc(max_slots, sizeof(*vals));
+	keys = calloc(max_slots, sizeof(*keys));
+	zeros = calloc(max_slots, sizeof(*zeros));
+
+	while (!bpf_map_get_next_key(fd, &lookup_key, &next_key)) {
+		bpf_map_lookup_elem(fd, &next_key, &val);
+		vals[next_key] = val;
+		lookup_key = next_key;
+		keys[i++] = next_key;
+	}
+
+	if (env.milliseconds)
+		print_log2_hist(vals, max_slots, "msecs");
+	else
+		print_log2_hist(vals, max_slots, "usecs");
+
+	err = bpf_map_update_batch(fd, keys, zeros, &i, &opts);
+	if (err < 0) {
+		fprintf(stderr, "clear map failed: %s\n", strerror(errno));
+		exit(1);
+	}
+
+	free(vals);
+	free(keys);
+	free(zeros);
+}
+
+static int pid_t_cmp(const void *n1, const void *n2)
+{
+	const struct dict_node *dn1 = n1, *dn2 = n2;
+
+	if (dn1->k == dn2->k)
+		return 0;
+	return dn1->k > dn2->k ? 1 : -1;
+}
+
+static void pid_to_comm(pid_t pid, char *comm)
+{
+	char fname[64];
+	FILE *f;
+
+	snprintf(fname, 64, "/proc/%d/comm", pid);
+	f = fopen(fname, "r");
+	if (!f || fscanf(f, "%s\n", comm) < 0)
+		*comm = '\0';
+}
+
+static void free_per_task_slots(void *nodep)
+{
+	struct dict_node *dn = nodep;
+
+	free(dn->v);
+}
+
+static void __print_per_task_log2_hist(const struct dict_node *dn)
+{
+	int max_slots = 65;
+	char comm[16];
+	pid_t pid;
+
+	pid = (intptr_t)dn->k;
+	pid_to_comm(pid, comm);
+	if (env.per_process)
+		printf("\npid = %d %s\n", pid, comm);
+	else
+		printf("\ntid = %d %s\n", pid, comm);
+	if (env.milliseconds)
+		print_log2_hist(dn->v, max_slots, "msecs");
+	else
+		print_log2_hist(dn->v, max_slots, "usecs");
+}
+
+DICT_ITER_FUNC(__print_per_task_log2_hist);
+
+static void print_per_task_log2_hist(int fd, int pid_max)
+{
+	struct dict d = { .root = NULL, .compar = pid_t_cmp };
+	struct piddata *lookup_key, *next_key, keys[pid_max];
+	DECLARE_LIBBPF_OPTS(bpf_map_batch_opts, opts,
```

Comment:
map batch operations are pretty recent addition, you'll be significantly limiting range of supported kernel. Is batching really needed, especially if you do one histogram as one map entry (instead of slot as map entry)?

Reply:
Oh, I didn't consider the issue of version restriction at that time, just to try the batch interface, I've fix it.